File: src/main/scala/org/rutiger/theatre/actors/exercise/one/Gimli.scala
package org.rutiger.theatre.actors.exercise.one

import akka.actor.{Actor, ActorLogging}
import org.rutiger.theatre.App.Attack
import org.rutiger.theatre.actors.exercise.Companion
import org.rutiger.theatre.actors.exercise.one.Gimli.Swung


class Gimli extends Actor with ActorLogging with Companion {
  private val MAX_SWINGING = 3

  override def receive: Receive = {
    case Attack => {
      log.info("Swinging my axe")
      sender() ! Swung(attackWith(MAX_SWINGING))
    }
    case _ => {
      log.warning("Me dont understand")
    }
  }
}

object Gimli {
  final case class Swung(axe: Int)
}

File: src/main/java/org/rutiger/theatre/actors/exercise/one/Gimli.java
package org.rutiger.theatre.actors.exercise.one;

import akka.actor.AbstractActor;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import org.rutiger.theatre.Messages;
import org.rutiger.theatre.actors.exercise.Companion;

public class Gimli extends AbstractActor implements Companion {
    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);
    private Integer counter = 0;

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Messages.Attack.class, s -> {
                    int axeSwung= attackWith(3);
                    log.info("swung axe {} times", axeSwung);
                    getSender().tell(new Messages.Shot(axeSwung), getSelf());
                })
                .match(Messages.Killed.class, killed -> {
                    counter += killed.orcs();
                })
                .match(Messages.EndBattle.class, any -> {
                    log.info("Hey i killed {} orcs", counter);
                    getContext().stop(getSelf());
                })
                .matchAny(o -> log.info("received unknown message"))
                .build();
    }

    public static Props props() {
        return Props.create(Gimli.class);
    }
}
